naga/src/proc/constant_evaluator.rs

```diff
@@ -1022,6 +1038,14 @@ impl<'a> ConstantEvaluator<'a> {
 
                 Expression::Compose { ty, components }
             }
+            Expression::Splat { size, value } => {
+                let value_span = self.expressions.get_span(value);
+                let cast_value = self.cast(value, target, value_span)?;
+                Expression::Splat {
+                    size,
+                    value: cast_value,
+                }
+            }
```

Comment:
It seems simple enough to also handle `ZeroValue` removing the need for the new `eval_zero_value` fn.

Reply:
At the moment, `eval_zero_value_impl` handles arrays, including nested arrays. Supporting those cases in `cast` would entail walking all the way into the type, adjusting the leaf scalar, and then rebuilding new types on the way back out. I think we will have to do that eventually to support array casting, but I would rather not take it on as part of this PR.